plugin.video.youtube/resources/lib/youtube_plugin/youtube/client/youtube.py

```diff
@@ -1443,167 +1391,386 @@ def search(self,
+                    output[_output[0]].update(_output[1])
+            else:
+                threads['balance'].clear()
 
-                    response.encoding = 'utf-8'
-                    xml_data = to_unicode(response.content)
-                    xml_data = xml_data.replace('\n', '')
-                    if do_encode:
-                        xml_data = to_str(xml_data)
-
-                    root = ET.fromstring(xml_data)
-                    _result['items'].extend([{
-                        'id': entry.find('yt:videoId', ns).text,
-                        'title': entry.find('media:group/media:title', ns).text,
-                        'channel': entry.find('atom:author/atom:name', ns).text,
-                        'published': entry.find('atom:published', ns).text,
-                    } for entry in root.findall('atom:entry', ns)])
-
-                # sorting by publish date
-                def _sort_by_date_time(item):
-                    return datetime_parser.since_epoch(
-                        datetime_parser.strptime(item['published'])
-                    )
+            thread = threading.current_thread()
+            threads['available'].release()
+            if dynamic:
+                threads['pool_counts'][pool_id] -= 1
+            threads['pool_counts']['all'] -= 1
+            threads['current'].discard(thread)
 
-                _result['items'].sort(reverse=True, key=_sort_by_date_time)
+        try:
+            num_cores = cpu_count() or 1
+        except NotImplementedError:
+            num_cores = 1
+        max_threads = min(32, 2 * (num_cores + 4))
+        threads = {
+            'max': max_threads,
+            'available': threading.Semaphore(max_threads),
+            'current': set(),
+            'pool_counts': {
+                'all': 0,
+            },
+            'balance': threading.Event(),
+        }
+        payloads = [
+            {
+                'pool_id': 1,
+                'kwargs': True,
+                'output': channel_ids,
+                'worker': _get_channels,
+                'threads': threads,
+                'limit': 1,
+                'dynamic': False,
+                'input_wait': None,
+            },
+        ] if logged_in else []
+        payloads.extend((
+            {
+                'pool_id': 2,
+                'kwargs': channel_ids,
+                'output': feeds,
+                'worker': _get_feed_cache,
+                'threads': threads,
+                'limit': 1,
+                'dynamic': False,
+                'input_wait': threading.Lock(),
+            },
+            {
+                'pool_id': 3,
+                'kwargs': channel_ids,
+                'output': feeds,
+                'worker': _get_feed,
+                'threads': threads,
+                'limit': None,
+                'dynamic': True,
+                'input_wait': threading.Lock(),
+            },
+        ))
+        while 1:
```

Comment:
Why the need for a `while 1`? It seems like this might cause a lot of threads to spawn?

Reply:
The number of threads that can be created is defined in line 1667-1669

This is checked in lines 1726-1734 to allow threads in the pools to be re-balanced (i.e. to be shutdown/repurposed and released allowing new threads to be added to the pool if needed)

The loop execution is also being paused if the number of threads exceeds the max allowable through the use of a threading.Semaphore object (created line 1670, acquired line 1744, and released by the various threads that are spawned on line 1657).

This means that the creation of new threads in the loop is always constrained, but still allows for a variable amount of threads to be monitored within the loop, spawning new threads if required, but also pausing thread execution if it has nothing to do, skipping idle threads while waiting for input from other threads, and releasing threads if their job has finished.